File: Python/house-robber.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    # @param num, a list of integer
    # @return an integer
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        last, now = 0, 0
        for i in nums:
            last, now = now, max(last + i, now)
        return now

File: C++/house-robber.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int rob(vector<int>& nums) {
        int last = 0, result = 0;
        for (const auto& i : nums) {
            auto tmp = result;
            result = max(last + i, result);
            last = tmp;
        }
        return result;
    }
};
